nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoCmisService.java

```diff
@@ -853,21 +863,46 @@ public AllowableActions getAllowableActions(String repositoryId, String objectId
     public ContentStream getContentStream(String repositoryId, String objectId, String streamId, BigInteger offset,
             BigInteger length, ExtensionsData extension) {
         // TODO offset, length
+        CallContext callContext = getCallContext();
+        ContentStream cs = null;
+        HttpServletRequest request = (HttpServletRequest) callContext.get(CallContext.HTTP_SERVLET_REQUEST);
         if (streamId == null) {
             DocumentModel doc = getDocumentModel(objectId);
-            HttpServletRequest request = (HttpServletRequest) getCallContext().get(CallContext.HTTP_SERVLET_REQUEST);
-            ContentStream cs = NuxeoPropertyData.getContentStream(doc, request);
-            if (cs != null) {
-                return cs;
+            cs = NuxeoPropertyData.getContentStream(doc, request);
+            if (cs == null) {
+                throw new CmisConstraintException("No content stream: " + objectId);
+            }
+        } else {
+            String renditionName = streamId.replaceAll("^" + REND_STREAM_RENDITION_PREFIX, "");
+            cs = getRenditionServiceStream(objectId, renditionName);
+            if (cs == null) {
+                throw new CmisInvalidArgumentException("Invalid stream id: " + streamId);
+            }
+        }
+        if (cs instanceof NuxeoContentStream) {
+            NuxeoContentStream ncs = (NuxeoContentStream) cs;
+            Blob blob = ncs.blob;
+            String digestAlgorithm = blob.getDigestAlgorithm();
+            if (AbstractBinaryManager.MD5_DIGEST.equals(digestAlgorithm) &&
+                    NuxeoContentStream.hasWantDigestRequestHeader(request,
+                            NuxeoContentStream.CONTENT_MD5_DIGEST_ALGORITHM)) {
+                setResponseHeader("Content-MD5", callContext, blob);
+            }
+            if (NuxeoContentStream.hasWantDigestRequestHeader(request, blob.getDigestAlgorithm())) {
+                setResponseHeader("Digest", callContext, blob);
             }
-            throw new CmisConstraintException("No content stream: " + objectId);
         }
-        String renditionName = streamId.replaceAll("^" + REND_STREAM_RENDITION_PREFIX, "");
-        ContentStream cs = getRenditionServiceStream(objectId, renditionName);
-        if (cs != null) {
-            return cs;
+        return cs;
+    }
+
+    protected void setResponseHeader(String headerName, CallContext callContext, Blob blob) {
+        try {
+            String digest = NuxeoPropertyData.transcodeHexToBase64(blob.getDigest());
+            HttpServletResponse response = (HttpServletResponse) callContext.get(CallContext.HTTP_SERVLET_RESPONSE);
+            String digestPrefix = "Digest".equals(headerName) ? blob.getDigestAlgorithm() + "=" : "";
+            response.setHeader(headerName, digestPrefix + digest);
+        } catch (DecoderException ignored) {
```

Comment:
Don't swallow exceptions. Anyway, don't deal with `DecoderException` here, but in the `transcodeHexToBase64` when you can rethrow as a `RuntimeException` as it's a "shouldn't happen" case.